File: mediaengines/helix/HxPlayer.cs
/* -*- Mode: csharp; tab-width: 4; c-basic-offset: 4; indent-tabs-mode: t -*- */
/***************************************************************************
 *  HxPlayer.cs
 *
 *  Copyright (C) 2005 Novell
 *  Written by Aaron Bockover (aaron@aaronbock.net)
 ****************************************************************************/

/*  THIS FILE IS LICENSED UNDER THE MIT LICENSE AS OUTLINED IMMEDIATELY BELOW: 
 *
 *  Permission is hereby granted, free of charge, to any person obtaining a
 *  copy of this software and associated documentation files (the "Software"),  
 *  to deal in the Software without restriction, including without limitation  
 *  the rights to use, copy, modify, merge, publish, distribute, sublicense,  
 *  and/or sell copies of the Software, and to permit persons to whom the  
 *  Software is furnished to do so, subject to the following conditions:
 *
 *  The above copyright notice and this permission notice shall be included in 
 *  all copies or substantial portions of the Software.
 *
 *  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR 
 *  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, 
 *  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE 
 *  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER 
 *  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING 
 *  FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER 
 *  DEALINGS IN THE SOFTWARE.
 */
 
using System;
using System.Runtime.InteropServices;

namespace Helix
{
	public class HxPlayer : IDisposable
	{		
		public const ushort PumpEventDelay = 75;
	
		public event ErrorOccurredHandler ErrorOccurred;
		public event LengthChangedHandler LengthChanged;
		public event VolumeChangedHandler VolumeChanged;
		public event ContentStateChangedHandler ContentStateChanged;
		public event StatusChangedHandler StatusChanged;
		public event ContentConcludedHandler ContentConcluded;
		public event MuteChangedHandler MuteChanged;
		public event BufferingHandler Buffering;
	
		private int token;
		private ClientEngineCallbacks callbacks;
	
		private uint length;
		private ushort volume;
		private int bandwidth;
		private string title;
		private string status;
		private HxContentState state;
		private bool muted;
		private string uri;
		
		private bool shouldIterate = false;
		
		public HxPlayer()
		{
			callbacks = new ClientEngineCallbacks();
			callbacks.OnVisualStateChanged = OnVisualStateChanged;
			callbacks.OnIdealSizeChanged = OnIdealSizeChanged;
			callbacks.OnLengthChanged = OnLengthChanged;
			callbacks.OnTitleChanged = OnTitleChanged;
			callbacks.OnGroupsChanged = OnGroupsChanged;
			callbacks.OnGroupStarted = OnGroupStarted;
			callbacks.OnContacting = OnContacting;
			callbacks.OnBuffering = OnBuffering;
			callbacks.OnContentStateChanged = OnContentStateChanged;
			callbacks.OnContentConcluded = OnContentConcluded;
			callbacks.OnStatusChanged = OnStatusChanged;
			callbacks.OnVolumeChanged = OnVolumeChanged;
			callbacks.OnMuteChanged = OnMuteChanged;
			callbacks.OnClipBandwidthChanged = OnClipBandwidthChanged;
			callbacks.OnErrorOccurred = OnErrorOccurred;
			callbacks.GoToURL = GoToURL;
			callbacks.RequestAuthentication = RequestAuthentication;
			callbacks.RequestUpgrade = RequestUpgrade;
			callbacks.HasComponent = HasComponent;
			callbacks.PrivateCallback1 = PrivateCallback;
			callbacks.PrivateCallback2 = PrivateCallback;
			
			IntPtr callbacksRaw = Marshal.AllocHGlobal(Marshal.SizeOf(callbacks));
			Marshal.StructureToPtr(callbacks, callbacksRaw, false);
				
			try {
				if(!HxUnmanaged.ClientPlayerCreate(out token, IntPtr.Zero, 
					IntPtr.Zero, callbacksRaw))
					throw new ApplicationException("Couldn't create player");
			} catch(NullReferenceException e) {
				throw new ApplicationException(
					"Couldn't create player: No HELIX_LIBS?");
			}
		}

		~HxPlayer()
		{
			Dispose();
		}

		public void Dispose()
		{
			HxUnmanaged.ClientPlayerClose(token);
			GC.SuppressFinalize(this);
		}
		
		public bool OpenUri(string uri)
		{
			shouldIterate = HxUnmanaged.ClientPlayerOpenURL(token, uri,
				null);
				
			if(shouldIterate)
				this.uri = uri;
				
			return shouldIterate;
		}

		public void Play()
		{
			HxUnmanaged.ClientPlayerPlay(token);
		}

		public void Pause()
		{
			HxUnmanaged.ClientPlayerPause(token);
		}
		
		public void Stop()
		{
			HxUnmanaged.ClientPlayerStop(token);
		}

		public void Iterate()
		{
			HxUnmanaged.ClientEngineProcessXEvent(IntPtr.Zero);
		}
		
		// Default Callbacks
		
		private void OnErrorOccurred(IntPtr player, uint hxCode, uint userCode, 
			IntPtr pErrorString, IntPtr pUserString, IntPtr pMoreInfoURL)
		{	
			ErrorOccurredHandler handler = ErrorOccurred;
			if(handler != null) {			
				string errorString = Marshal.PtrToStringAnsi(pErrorString);
				string userString = Marshal.PtrToStringAnsi(pUserString);
				string moreInfoUrl = Marshal.PtrToStringAnsi(pMoreInfoURL);
				
				ErrorOccurredArgs args = new ErrorOccurredArgs();
				args.Player = this;
				args.Error = errorString;
				args.UserError = userString;
				args.MoreInfoUrl = moreInfoUrl;
				
				handler(this, args);
			}			
		}

		private void OnContentStateChanged(IntPtr player, 
			HxContentState oldContentState, HxContentState newContentState)
		{
			state = newContentState;

			ContentStateChangedHandler handler = ContentStateChanged;
			if(handler != null) {
				ContentStateChangedArgs args = new ContentStateChangedArgs();
				args.Player = this;
				args.NewState = newContentState;
				args.OldState = oldContentState;
				
				handler(this, args);
			}
		}

		private void OnStatusChanged(IntPtr player, IntPtr pStatus)
		{
			status = Marshal.PtrToStringAnsi(pStatus);
			
			StatusChangedHandler handler = StatusChanged;
			if(handler != null) {
				StatusChangedArgs args = new StatusChangedArgs();
				args.Player = this;
				args.Status = status;
				
				handler(this, args);
			}
		}

		private void OnMuteChanged(IntPtr player, bool hasMuted)
		{
			muted = hasMuted;
			
			MuteChangedHandler handler = MuteChanged;
			if(handler != null) {
				MuteChangedArgs args = new MuteChangedArgs();
				args.Player = this;
				args.Muted = muted;
				
				handler(this, args);
			}
		}

		private void OnContentConcluded(IntPtr player)
		{
			ContentConcludedHandler handler = ContentConcluded;
			if(handler != null) {
				HxPlayerArgs args = new HxPlayerArgs();
				args.Player = this;
				
				handler(this, args);
			}
			
			shouldIterate = false;
		}

		private void OnTitleChanged(IntPtr player, IntPtr pTitle)
		{
			title = Marshal.PtrToStringAnsi(pTitle);
		}

		private void OnLengthChanged(IntPtr player, uint length)
		{
			this.length = length;
				
			LengthChangedHandler handler = LengthChanged;
			if(handler != null) {
				LengthChangedArgs args = new LengthChangedArgs();
				args.Player = this;
				args.Length = length;
				
				handler(this, args);
			}		
		}

		private void OnVolumeChanged(IntPtr player, UInt16 volume)
		{
			this.volume = volume;
				
			VolumeChangedHandler handler = VolumeChanged;
			if(handler != null) {
				VolumeChangedArgs args = new VolumeChangedArgs();
				args.Player = this;
				args.Volume = volume;
				
				handler(this, args);
			}		
		}

		private void OnClipBandwidthChanged(IntPtr player, int clipBandwidth)
		{
			bandwidth = clipBandwidth;
		}

		private void OnBuffering(IntPtr player, HxBufferReason bufferingReason, 
			UInt16 bufferingPercent)
		{
			BufferingHandler handler = Buffering;
			if(handler != null) {
				BufferingArgs args = new BufferingArgs();
				args.Player = this;
				args.BufferingReason = bufferingReason;
				args.BufferingPercent = bufferingPercent;
				
				handler(this, args);
			}
		}

		// Not supported yet in these bindings

		private void OnVisualStateChanged(IntPtr player, bool hasVisualContent)
		{
		
		}

		private void OnContacting(IntPtr player, IntPtr pHostName)
		{
		
		}
		
		private void OnIdealSizeChanged(IntPtr player, int idealWidth, 
			int idealHeight)
		{
		
		}
		
		private void OnGroupsChanged(IntPtr player)
		{
		
		}

		private void OnGroupStarted(IntPtr player, UInt16 groupIndex)
		{
		
		}

		private bool GoToURL(IntPtr player, IntPtr url,  IntPtr target, 
			bool isPlayerUrl, bool isAutoActivated)
		{
			return false;
		}
		
		private bool HasComponent(IntPtr player, IntPtr componentName)
		{	
			return false;
		}

		private bool RequestUpgrade(IntPtr player, IntPtr pUrl, 
			uint numOfComponents, IntPtr componentNamesArr, bool isBlocking)
		{
			return false;
		}

		private bool RequestAuthentication(IntPtr player, IntPtr pServer, 
			IntPtr pRealm, bool isProxyServer)
		{
			return false;
		}
		
		private void PrivateCallback(IntPtr userInfo)
		{
			
		}
				
		public uint Length
		{
			get {
				return length;
			}
		}
		
		public ushort Volume
		{
			get {
				return volume;
			}
			
			set {
				HxUnmanaged.ClientPlayerSetVolume(token, value);
			}
		}
		
		public int Bandwidth
		{
			get {
				return bandwidth;
			}
		}
		
		public string Title
		{
			get {
				return title;
			}
		}
		
		public string Status
		{
			get {
				return status;
			}
		}
		
		public HxContentState State
		{
			get {
				return state;
			}
		}
		
		public string Uri
		{
			get {
				return uri;
			}
		}
		
		public bool Muted 
		{
			get {
				return muted;
			}
			
			set {
				HxUnmanaged.ClientPlayerMute(token, value);
			}
		}
		
		public uint Position
		{
			get {
				return HxUnmanaged.ClientPlayerGetPosition(token);
			}
			
			set {
				HxUnmanaged.ClientPlayerSetPosition(token, value);
			}
		}
	}
}

File: mediaengines/helix/HXPlayer.c
/***************************************************************************
 *  HXPlayer.c
 *
 *  Copyright (C) 2005 Novell
 *  Written by Aaron Bockover (aaron@aaronbock.net)
 ****************************************************************************/

/*  THIS FILE IS LICENSED UNDER THE MIT LICENSE AS OUTLINED IMMEDIATELY BELOW: 
 *
 *  Permission is hereby granted, free of charge, to any person obtaining a
 *  copy of this software and associated documentation files (the "Software"),  
 *  to deal in the Software without restriction, including without limitation  
 *  the rights to use, copy, modify, merge, publish, distribute, sublicense,  
 *  and/or sell copies of the Software, and to permit persons to whom the  
 *  Software is furnished to do so, subject to the following conditions:
 *
 *  The above copyright notice and this permission notice shall be included in 
 *  all copies or substantial portions of the Software.
 *
 *  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR 
 *  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, 
 *  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE 
 *  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER 
 *  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING 
 *  FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER 
 *  DEALINGS IN THE SOFTWARE.
 */
 
#include <stdlib.h>
#include <stdio.h>
#include <memory.h>

#include "HXPlayer.h"

/* Callback Execution Macros */

#define CBC(PLAYER_PTR, CALLBACK)              \
	HXPlayer *player = (HXPlayer *)PLAYER_PTR; \
	if(player->callbacks.CALLBACK != NULL)

#define RUN_CALLBACK_1(PLAYER_PTR, CALLBACK, A) \
	{ CBC(PLAYER_PTR, CALLBACK)(player->callbacks.CALLBACK)(A); } 

#define RUN_CALLBACK_2(PLAYER_PTR, CALLBACK, A, B) \
	{ CBC(PLAYER_PTR, CALLBACK)(player->callbacks.CALLBACK)(A, B); } 
	
#define RUN_CALLBACK_3(PLAYER_PTR, CALLBACK, A, B, C) \
	{ CBC(PLAYER_PTR, CALLBACK)(player->callbacks.CALLBACK)(A, B, C); } 
	
#define RUN_CALLBACK_6(PLAYER_PTR, CALLBACK, A, B, C, D, E, F) \
	{ CBC(PLAYER_PTR, CALLBACK)(player->callbacks.CALLBACK)(A, B, C, D, E, F); } 
	
/* Default Client Engine Callback Signatures */
	
static void OnVisualStateChanged(void* userInfo, bool hasVisualContent);
static void OnIdealSizeChanged(void* userInfo, SInt32 idealWidth, 
	SInt32 idealHeight);

static void OnLengthChanged(void* userInfo, UInt32 length);
static void OnTitleChanged(void* userInfo, const char* pTitle);
static void OnGroupsChanged(void* userInfo);
static void OnGroupStarted(void* userInfo, UInt16 groupIndex);
static void OnContacting(void* userInfo, const char* pHostName);

static void OnBuffering(void* userInfo, UInt32 bufferingReason, 
	UInt16 bufferPercent);

static void OnContentConcluded(void* userInfo);
static void OnContentStateChanged(void* userInfo, int oldContentState, 
	int newContentState);

static void OnStatusChanged(void* userInfo, const char* pStatus);
static void OnVolumeChanged(void* userInfo, UInt16 volume);
static void OnMuteChanged(void* userInfo, bool hasMuted);
static void OnClipBandwidthChanged(void* userInfo, SInt32 clipBandwidth);

static void OnErrorOccurred(void* userInfo, UInt32 hxCode, UInt32 userCode,
		const char* pErrorString, const char* pUserString,
		const char* pMoreInfoURL);

static bool GoToURL(void* userInfo, const char* pURL, const char* pTarget, 
	bool isPlayerURL, bool isAutoActivated);

static bool RequestAuthentication(void* userInfo, const char* pServer, 
	const char* pRealm, bool isProxyServer);

static bool RequestUpgrade(void* userInfo, const char* pUrl, 
	UInt32 numOfComponents, const char* componentNames[], bool isBlocking);
		   
static bool HasComponent(void* userInfo, const char* componentName);

/* Default Client Engine Callback Connection */

static const HXClientCallbacks hxPlayerCallbacks = {
	OnVisualStateChanged,
	OnIdealSizeChanged,
	OnLengthChanged,
	OnTitleChanged,
	OnGroupsChanged,
	OnGroupStarted,
	OnContacting,
	OnBuffering,
	OnContentStateChanged,
	OnContentConcluded,
	OnStatusChanged,
	OnVolumeChanged,
	OnMuteChanged,
	OnClipBandwidthChanged,
	OnErrorOccurred,
	GoToURL,
	RequestAuthentication,
	RequestUpgrade,
	HasComponent
};

/* Default Client Engine Callbacks */ 

static void
OnVisualStateChanged(void* userInfo, bool hasVisualContent)
{
	RUN_CALLBACK_2(userInfo, OnVisualStateChanged, userInfo, hasVisualContent);
}

static void
OnIdealSizeChanged(void* userInfo, SInt32 idealWidth, SInt32 idealHeight)
{
	RUN_CALLBACK_3(userInfo, OnIdealSizeChanged, userInfo, 
		idealWidth, idealHeight);
}

static void
OnLengthChanged(void* userInfo, UInt32 length)
{
	RUN_CALLBACK_2(userInfo, OnLengthChanged, userInfo, length);
}

static void
OnTitleChanged(void* userInfo, const char* pTitle)
{
	RUN_CALLBACK_2(userInfo, OnTitleChanged, userInfo, pTitle);
}

static void
OnGroupsChanged(void* userInfo)
{
	RUN_CALLBACK_1(userInfo, OnGroupsChanged, userInfo);
}

static void
OnGroupStarted(void* userInfo, UInt16 groupIndex)
{
	RUN_CALLBACK_2(userInfo, OnGroupStarted, userInfo, groupIndex);
}

static void
OnContacting(void* userInfo, const char* pHostName)
{
	RUN_CALLBACK_2(userInfo, OnContacting, userInfo, pHostName);
}

static void
OnBuffering(void* userInfo, UInt32 bufferingReason, UInt16 bufferPercent)
{
	RUN_CALLBACK_3(userInfo, OnBuffering, userInfo, bufferingReason,
		bufferPercent);
}

static void
OnContentConcluded(void* userInfo)
{
	RUN_CALLBACK_1(userInfo, OnContentConcluded, userInfo);
}

static void
OnContentStateChanged(void* userInfo, int oldContentState, int newContentState)
{
	RUN_CALLBACK_3(userInfo, OnContentStateChanged, userInfo, oldContentState,
		newContentState);
}

static void
OnStatusChanged(void* userInfo, const char* pStatus)
{
	RUN_CALLBACK_2(userInfo, OnStatusChanged, userInfo, pStatus);
}

static void
OnVolumeChanged(void* userInfo, UInt16 volume)
{
	RUN_CALLBACK_2(userInfo, OnVolumeChanged, userInfo, volume);
}

static void
OnMuteChanged(void* userInfo, bool hasMuted)
{
	RUN_CALLBACK_2(userInfo, OnMuteChanged, userInfo, hasMuted);
}

static void
OnClipBandwidthChanged(void* userInfo, SInt32 clipBandwidth)
{
	RUN_CALLBACK_2(userInfo, OnClipBandwidthChanged, userInfo, clipBandwidth);
}

static void
OnErrorOccurred(void* userInfo, UInt32 hxCode, UInt32 userCode,
		const char* pErrorString, const char* pUserString,
		const char* pMoreInfoURL)		
{
	RUN_CALLBACK_6(userInfo, OnErrorOccurred, userInfo, hxCode, userCode,
		pErrorString, pUserString, pMoreInfoURL);
}

bool
GoToURL(void* userInfo, const char* pURL, const char* pTarget, 
	bool isPlayerURL, bool isAutoActivated)
{
    return FALSE;
}

bool
RequestAuthentication(void* userInfo, const char* pServer, const char* pRealm,
		      bool isProxyServer)
{
    return FALSE;
}

bool
RequestUpgrade(void* userInfo, const char* pUrl, UInt32 numOfComponents,
	       const char* componentNames[], bool isBlocking)
{
    return FALSE;
}

bool
HasComponent(void* userInfo, const char* componentName)
{
    return FALSE;
}

/* HXPlayer Client Engine Core Wrapper */

HXPlayer *HXPlayerCreate()
{
	HXPlayer *player = (HXPlayer *)malloc(sizeof(HXPlayer));
	if(player == NULL)
		return NULL;
	
	player->handle = 0;
	player->state = 0;
	player->initialized = FALSE;
	
	memset(&player->callbacks, 0, sizeof(HXClientCallbacks));
	
	return player;
}

bool HXPlayerInit(HXPlayer *player)
{
	if(player == NULL)
		return FALSE;
	
	setenv("HELIX_LIBS", "/usr/lib/RealPlayer10", 0);
	
	if(ClientPlayerCreate(&player->handle, 0, player, &hxPlayerCallbacks))
		player->initialized = TRUE;
		
	return player->initialized;
}

void HXPlayerShutdown(HXPlayer *player)
{
	if(player == NULL)
		return;
	
	if(!player->initialized)
		return;
	
	ClientPlayerClose(player->handle);
	player->initialized = FALSE;
}
	
void HXPlayerFree(HXPlayer *player)
{
	if(player == NULL);
		return;
	
	if(player->initialized)
		HXPlayerShutdown(player);
	
	free(player);
}

int HXPlayerIterate(HXPlayer *player)
{
	if(player == NULL)
		return 0;
	
	usleep(20);
	ClientEngineProcessXEvent(NULL);
	return ClientPlayerGetContentState(player->handle);
}

HXClientPlayerToken HXPlayerGetHandle(HXPlayer *player)
{
	if(player == NULL)
		return NULL;
	
	return player->handle;
}

/* This is a hack... need to figure out why mono won't properly p/invoke
   into hxclient, but works fine here. */
   
unsigned int HXPlayerGetVolume(HXPlayer *player)
{
	if(player == NULL)
		return 0;
		
	return ClientPlayerGetVolume(player->handle);
}

void HXPlayerSetVolume(HXPlayer *player, unsigned int volume)
{
	if(player == NULL)
		return;
		
	ClientPlayerSetVolume(player->handle, volume);
}

int HXPlayerGetPosition(HXPlayer *player)
{
	if(player == NULL)
		return -1;
		
	return ClientPlayerGetPosition(player->handle);
}

void HXPlayerSetPosition(HXPlayer *player, int position)
{
	if(player == NULL)
		return;
		
	ClientPlayerSetPosition(player->handle, position);
}

void HXPlayerStop(HXPlayer *player)
{
	if(player == NULL)
		return;
		
	ClientPlayerStop(player->handle);
}

bool HXPlayerOpenUrl(HXPlayer *player, const char *url)
{
	return ClientPlayerOpenUrl(player->handle, url, NULL);
}

void HXPlayerPlay(HXPlayer *player)
{
	if(player == NULL)
		return;
		
	ClientPlayerPlay(player->handle);
}

void HXPlayerPause(HXPlayer *player)
{
	if(player == NULL)
		return;
		
	ClientPlayerPause(player->handle);
}

/* Client Engine Callback Registrations */

void HXPlayerRegisterOnVisualStateChangedCallback(HXPlayer *player, 
	HXOnVisualStateChangedProcPtr cb)
{
	if(player != NULL)
		player->callbacks.OnVisualStateChanged = cb;
}

void HXPlayerRegisterOnIdealSizeChangedCallback(HXPlayer *player, 
	HXOnIdealSizeChangedProcPtr cb)
{
	if(player != NULL)
		player->callbacks.OnIdealSizeChanged = cb;
}

void HXPlayerRegisterOnLengthChangedCallback(HXPlayer *player, 
	HXOnLengthChangedProcPtr cb)
{
	if(player != NULL)
		player->callbacks.OnLengthChanged = cb;
}

void HXPlayerRegisterOnTitleChangedCallback(HXPlayer *player, 
	HXOnTitleChangedProcPtr cb)
{
	if(player != NULL)
		player->callbacks.OnTitleChanged = cb;
}

void HXPlayerRegisterOnGroupsChangedCallback(HXPlayer *player, 
	HXOnGroupsChangedProcPtr cb)
{
	if(player != NULL)
		player->callbacks.OnGroupsChanged = cb;
}

void HXPlayerRegisterOnGroupStartedCallback(HXPlayer *player, 
	HXOnGroupStartedProcPtr cb)
{
	if(player != NULL)
		player->callbacks.OnGroupStarted = cb;
}

void HXPlayerRegisterOnContactingCallback(HXPlayer *player, 
	HXOnContactingProcPtr cb)
{
	if(player != NULL)
		player->callbacks.OnContacting = cb;
}

void HXPlayerRegisterOnBufferingCallback(HXPlayer *player, 
	HXOnBufferingProcPtr cb)
{
	if(player != NULL)
		player->callbacks.OnBuffering = cb;
}

void HXPlayerRegisterOnContentStateChangedCallback(HXPlayer *player, 
	HXOnContentStateChangedProcPtr cb)
{
	if(player != NULL)
		player->callbacks.OnContentStateChanged = cb;
}

void HXPlayerRegisterOnContentConcludedCallback(HXPlayer *player, 
	HXOnContentConcludedProcPtr cb)
{
	if(player != NULL)
		player->callbacks.OnContentConcluded = cb;
}

void HXPlayerRegisterOnStatusChangedCallback(HXPlayer *player, 
	HXOnStatusChangedProcPtr cb)
{
	if(player != NULL)
		player->callbacks.OnStatusChanged = cb;
}

void HXPlayerRegisterOnVolumeChangedCallback(HXPlayer *player, 
	HXOnVolumeChangedProcPtr cb)
{
	if(player != NULL)
		player->callbacks.OnVolumeChanged = cb;
}

void HXPlayerRegisterOnMuteChangedCallback(HXPlayer *player, 
	HXOnMuteChangedProcPtr cb)
{
	if(player != NULL)
		player->callbacks.OnMuteChanged = cb;
}

void HXPlayerRegisterOnClipBandwidthChangedCallback(HXPlayer *player, 
	HXOnClipBandwidthChangedProcPtr cb)
{
	if(player != NULL)
		player->callbacks.OnClipBandwidthChanged = cb;
}

void HXPlayerRegisterOnErrorOccurredCallback(HXPlayer *player, 
	HXOnErrorOccurredProcPtr cb)
{
	if(player != NULL)
		player->callbacks.OnErrorOccurred = cb;
}

void HXPlayerRegisterGoToURLCallback(HXPlayer *player, 
	HXGoToURLProcPtr cb)
{
	if(player != NULL)
		player->callbacks.GoToURL = cb;
}

void HXPlayerRegisterRequestAuthenticationCallback(HXPlayer *player, 
	HXRequestAuthenticationProcPtr cb)
{
	if(player != NULL)
		player->callbacks.RequestAuthentication = cb;
}

void HXPlayerRegisterRequestUpgradeCallback(HXPlayer *player, 
	HXRequestUpgradeProcPtr cb)
{
	if(player != NULL)
		player->callbacks.RequestUpgrade = cb;
}

void HXPlayerRegisterHasComponentCallback(HXPlayer *player, 
	HXHasComponentProcPtr cb)
{
	if(player != NULL)
		player->callbacks.HasComponent = cb;
}

void RegisterGeneric(HXPlayer *player, void *ptr)
{
	player->callbacks.HasComponent = ptr;	
}

void ass()
{
	RegisterGeneric(NULL, RegisterGeneric);
}
